### ISSUE TITLE
Critical Feature Request :: SubDNA

### ISSUE THREAD

**Issue by burmestr** [HashLips/hashlips_art_engine#46]
Here is the problem that I cannot solve myself, please help.
I actually need some kind of SubDNA to make it working this way.

Example:

- Background
- Body
- Clothing
- Banner
- Word1
- Word2
- Word3
- Object

I'm fine if bodies, backgrounds, clothings will mix and repeat sometime, but I want the script to include ONLY unique combinations of Word1+Word2+Word3. And of course there are more than enough words inside to meet the Edition size. 

Right now the script may pick the same Word1+Word2+Word3 since the background or clothing will be different for DNA. 

As far as I understand there must be additional "DNA Exists" check for the specified folders to ensure it won't repeat the combinations, but I can't figure out how to make it myself. I already tried creating one edition with just unique words and another edition with just images. I know how to merge PNGs, but I have no idea how to merge metadata (json) then. 

Please help!

**Comment by burmestr** [comment]
Alternatively, if this subDNA idea is complicated for the Art Engine, the solution for my other feature request could be a workaround ( See issue #47 ) because I could make 2-3 editions of trait names and name it using traits, and then use the results as traits for the final generation and transfer meta data this way, but it's quite dummy with its own cons.

**Comment by Ahmeddh** [comment]
this has been solved in v1.0.4 I guess, where you have an array of layers configuration, in your case you simply an array of three items each have one of the words layer, see the below layers configuration example.
`const layerConfigurations = [
  {
    growEditionSizeTo: 20,
    layersOrder: [
      { name: "Background" },
      { name: "Body" },
      { name: "Clothing" },
      { name: "Banner" },
      { name: "Word1" },
      { name: "Object" },
    ],
  },
  {
    growEditionSizeTo: 40,
    layersOrder: [
      { name: "Background" },
      { name: "Body" },
      { name: "Clothing" },
      { name: "Banner" },
      { name: "Word2" },
      { name: "Object" },
    ],
  },
  {
    growEditionSizeTo: 50,
    layersOrder: [
      { name: "Background" },
      { name: "Body" },
      { name: "Clothing" },
      { name: "Banner" },
      { name: "Word3" },
      { name: "Object" },
    ],
  },
];`

In the above example the first set will have 20 items, the one after will also have 20 items, and the final set will have 10 items, notice the growEditionSizeTo (each object has to be bigger than the one before it)

I hope it solves your issue

**Comment by burmestr** [comment]
The issue I've described has nothing to do with layer configurations

I don't need Word1 with one part of edition and Word2 with another, I need the script to use only unique combinations of Word1+Word2+Word3!

**Comment by burmestr** [comment]
Imagine that you create NFT collection of cars, and they have plate numbers, you've made layers for every digit/letter of the plate number. The script mixed layers to make unique DNA, if your edition is large enough, you'll get repeating car plate numbers, so you'll have 7 different cars with identical car plate numbers etc, though mathematically you had enough variatons to generate a billion of car plate numbers. But right now the script will be happy to use the same car plate combinations because it will use differnt spoiler color or background and still get unique DNA. That's why there must be SubDNA, to group layer groups that must generate unique variations. Otherwise you have to split the project in parts, generate 10-50-100k of each part, and then think how to merge them all together and merge metadata..

**Comment by jasan-s** [comment]
I may work on this weekend and submit a PR

**Comment by burmestr** [comment]
@jasan-s that would be amazing! Right now I'm trying a dummy workaround generating 2 editions and then merging final PNGs together as well as metadata, so in the end the script will be used merge just 2 folders with 50k PNGs each without a chance of repeating anything. But I need to make sure that it will mege 1.png from Folder1 with 1.png from Folder2, otherwise I will lose track of meta data :-) Please keep everyone posted about your progress! Appreciated!

**Comment by burmestr** [comment]
OK so I've ran the script separately for the part that must be unique in my project and got 50k+ unique results. Fortunately I've managed to extract metadata and rename all 50k PNG files according to its metadata. Now I added this pack of PNGs as an additional layer to my primary project, all 50k files had equal rarity weight (1). The edition size was 50k, dnatorrance was 50k, stopped the script manually at 1500 editions, checked metadata, it already used some of the layers twice, so it doesn't care that there are 50k layers inside the folder. 
So I really wish I could add some special symbol to this layer so the script would not allow to pick the same file more than once from that folder. But of course it would be better (and would make more sense) if I could select a group of layers that must generate unique and only unique variations within the whole project...
